validator/client/runner.go

```diff
@@ -24,123 +24,171 @@ import (
+					if slotCancel != nil {
+						slotCancel()
+					}
+					return
+				}
+			}
+
+			slotCtx, slotCancel = context.WithDeadline(ctx, v.SlotDeadline(slot))
+			slotCtx, slotSpan = trace.StartSpan(slotCtx, "validator.processSlot")
+			slotSpan.AddAttributes(trace.Int64Attribute("slot", int64(slot))) // lint:ignore uintcast -- This conversion is OK for tracing.
 
-			// call push proposer setting at the start of each epoch to account for the following edge case:
-			// proposer is activated at the start of epoch and tries to propose immediately
 			if slots.IsEpochStart(slot) {
-				go func() {
-					// deadline set for 1 epoch from call to not overlap.
-					epochDeadline := v.SlotDeadline(slot + params.BeaconConfig().SlotsPerEpoch - 1)
-					if err := v.PushProposerSettings(ctx, km, slot, epochDeadline); err != nil {
-						log.WithError(err).Warn("Failed to update proposer settings")
-					}
-				}()
+				updateDutiesNeeded = true
+				// Update proposer settings at the start of each epoch because new validators could have become active
+				pushProposerSettingsChan <- slot
 			}
 
-			// Start fetching domain data for the next epoch.
-			if slots.IsEpochEnd(slot) {
-				go v.UpdateDomainDataCaches(ctx, slot+1)
+			if updateDutiesNeeded {
+				updateDutiesNeeded = false
+				if err = v.UpdateDuties(ctx, slot); err != nil {
+					handleUpdateDutiesError(err, slot)
+					updateDutiesNeeded = true
+					slotCancel()
+					slotSpan.End()
+					continue
+				}
 			}
 
 			var wg sync.WaitGroup
-
-			allRoles, err := v.RolesAt(ctx, slot)
+			roles, err := v.RolesAt(slotCtx, slot)
 			if err != nil {
 				log.WithError(err).Error("Could not get validator roles")
-				cancel()
-				span.End()
+				slotCancel()
+				slotSpan.End()
 				continue
 			}
-			performRoles(slotCtx, allRoles, v, slot, &wg, span)
-		case isHealthyAgain := <-healthTracker.HealthUpdates():
-			if isHealthyAgain {
-				headSlot, err = initializeValidatorAndGetHeadSlot(ctx, v)
-				if err != nil {
-					log.WithError(err).Error("Failed to re initialize validator and get head slot")
-					continue
+			performRoles(slotCtx, roles, v, slot, &wg, slotSpan)
+
+			// Start fetching domain data for the next epoch.
+			if slots.IsEpochEnd(slot) {
+				go v.UpdateDomainDataCaches(ctx, slot+1)
+			}
+		case slot := <-v.LastSecondOfSlot():
+			nodeIsHealthyCurr := v.HealthTracker().IsHealthy()
+			if nodeIsHealthyCurr {
+				if !nodeIsHealthyPrev {
+					pushProposerSettingsChan <- slot
 				}
-				if err := v.UpdateDuties(ctx, headSlot); err != nil {
-					handleAssignmentError(err, headSlot)
-					continue
+				// In case event stream died
+				if !v.EventStreamIsRunning() {
+					log.Info("Event stream reconnecting...")
+					go v.StartEventStream(ctx, event.DefaultEventTopics, eventsChan)
+				}
+			} else if features.Get().EnableBeaconRESTApi {
+				v.ChangeHost()
+				initializationNeeded = true
+				updateDutiesNeeded = true
```

Comment:
One question that might come up is "why do we update duties but don't push proposer settings here?"

The answer is because they would have been pushed immediately, without waiting for the beacon node to become healthy, whereas update duties will not be called before the node comes back up.